### ISSUE TITLE
GRPO training the Qwen2.5-7B model, after 30 steps, NaN appeares in grad_norm, but other loss is normal.

### ISSUE THREAD

**Issue by zyf8818** [verl-project/verl#2708]
When training the Qwen2.5-7B model, after 30 steps, NaN appeares in grad_norm, but other loss-related indicators are normal. I made some cuts to "roll_out", decoding it first and then re-encoding it back.

<img width="690" height="300" alt="Image" src="https://github.com/user-attachments/assets/a489bfab-2f4f-434f-b7fc-29c751478ec4" />
<img width="690" height="300" alt="Image" src="https://github.com/user-attachments/assets/96ead098-0905-489b-9859-84f16e808c61" />
<img width="690" height="300" alt="Image" src="https://github.com/user-attachments/assets/e40ff2e2-4f84-4351-9141-ed07ff17050d" />
<img width="689" height="300" alt="Image" src="https://github.com/user-attachments/assets/221f06ec-a2c4-420a-bc8b-ae2ebdd41e8a" />
<img width="690" height="300" alt="Image" src="https://github.com/user-attachments/assets/0bd581c6-1149-4047-a32a-9dc6c5ff9fac" />
<img width="690" height="300" alt="Image" src="https://github.com/user-attachments/assets/ae921189-ba67-4f65-9fc2-7fed764dd81f" />
<img width="690" height="300" alt="Image" src="https://github.com/user-attachments/assets/f0e1817f-6821-4b50-b7e6-d39325becb6e" />

It is trained on 4 H20.
May I ask if any of you have encountered similar problems?

**Comment by tdano1** [comment]
What dataset are you using? My pg loss surges after 50 steps but the grad norm remains normal..

**Comment by zyf8818** [comment]
> What dataset are you using? My pg loss surges after 50 steps but the grad norm remains normal..

I used my own dataset. I have located the problem, the roll_out is None and there is no end token.

**Comment by zyf8818** [comment]
The problem is caused by the error in the roll_outs. some roll_outs are empty but there is no end token.

**Comment by threegold116** [comment]
> The problem is caused by the error in the roll_outs. some roll_outs are empty but there is no end token.该问题是由 roll_outs 中的错误引起的。有些 roll_outs 是空的，但没有结束令牌。

what means "empty"? no tokens?

**Comment by zyf8818** [comment]
> > The problem is caused by the error in the roll_outs. some roll_outs are empty but there is no end token.该问题是由 roll_outs 中的错误引起的。有些 roll_outs 是空的，但没有结束令牌。
> 
> what means "empty"? no tokens?

Yes, there is no token in some roll_out

**Comment by lebronjamesking** [comment]
> > > The problem is caused by the error in the roll_outs. some roll_outs are empty but there is no end token.该问题是由 roll_outs 中的错误引起的。有些 roll_outs 是空的，但没有结束令牌。
> > 
> > 
> > what means "empty"? no tokens?
> 
> Yes, there is no token in some roll_out

I think so, some rollouts may be padded with empty response. However, if we use latest feature over_sample_rate (fixing the problem of some rollouts hang on whilst generating extra long response may even not stop generating) introduced in sglang_rollout.py, grad_norm nan is plausible. See details here: https://github.com/volcengine/verl/commit/ea885f32f04d86c3a81de18083db7eef0d781421#commitcomment-164178792

**Comment by wutong4012** [comment]
I encountered a similar problem and spent a long time troubleshooting it without any clues.

However, I observed that when calculating `old_log_probs `and `ref_log_prob`, no NaNs were generated.

However, when updating the actor, NaNs were generated. While they were not generated the first time, they did appear after a `loss.backward()` call.

**Comment by zyf8818** [comment]
> I encountered a similar problem and spent a long time troubleshooting it without any clues.
> 
> However, I observed that when calculating `old_log_probs `and `ref_log_prob`, no NaNs were generated.
> 
> However, when updating the actor, NaNs were generated. While they were not generated the first time, they did appear after a `loss.backward()` call.

I ended up having an empty content because I modified the samplings, but the end token was missing.

**Comment by lebronjamesking** [comment]
> > I encountered a similar problem and spent a long time troubleshooting it without any clues.
> > However, I observed that when calculating `old_log_probs `and `ref_log_prob`, no NaNs were generated.
> > However, when updating the actor, NaNs were generated. While they were not generated the first time, they did appear after a `loss.backward()` call.
> 
> I ended up having an empty content because I modified the samplings, but the end token was missing.

For example, in qwen series, what token is end token? <|endoftext|>

**Comment by zyf8818** [comment]
> > > I encountered a similar problem and spent a long time troubleshooting it without any clues.
> > > However, I observed that when calculating `old_log_probs `and `ref_log_prob`, no NaNs were generated.
> > > However, when updating the actor, NaNs were generated. While they were not generated the first time, they did appear after a `loss.backward()` call.
> > 
> > 
> > I ended up having an empty content because I modified the samplings, but the end token was missing.
> 
> For example, in qwen series, what token is end token? <|endoftext|>

this token <|im_end|>
When I modified the sample, the extracted content was empty, but this token was not added at the end.

**Comment by lebronjamesking** [comment]
> > > > I encountered a similar problem and spent a long time troubleshooting it without any clues.
> > > > However, I observed that when calculating `old_log_probs `and `ref_log_prob`, no NaNs were generated.
> > > > However, when updating the actor, NaNs were generated. While they were not generated the first time, they did appear after a `loss.backward()` call.
> > > 
> > > 
> > > I ended up having an empty content because I modified the samplings, but the end token was missing.
> > 
> > 
> > For example, in qwen series, what token is end token? <|endoftext|>
> 
> this token <|im_end|> When I modified the sample, the extracted content was empty, but this token was not added at the end.

Cool. Do you think for example in the sglang fashion and in sglang_rollout.py we can add 

`        request_sampling_params.update ({'stop':['<|endoftext|>','<|im_end|>'],'no_stop_trim':True})
`
at function _async_rollout_a_request() to avoid special tokens causing grad_norm nan.

**Comment by Kirrito-k423** [comment]
### 中文
定时清理检查发现：此 issue 最近一次非当前运行账号的公开回复或引用时间是 `2025-08-22T07:25:02Z`（comment，来自 `lebronjamesking`），相对于本轮运行日期 `2026-07-12` 已超过 6 个月。因此按社区 issue 定时清理标准关闭。

已检查证据：Latest non-self public reply/reference: `2025-08-22T07:25:02Z` (comment by `lebronjamesking`).
Comments API: https://api.github.com/repos/verl-project/verl/issues/2708/comments
Timeline API: https://api.github.com/repos/verl-project/verl/issues/2708/timeline

如果作者或社区成员仍能提供新的复现、证据或需求背景，请在此 issue 下继续评论请求重开，或创建新的 issue 并链接到这里。

### English
Scheduled cleanup check found that the latest public reply or reference by an account other than the current automation account was `2025-08-22T07:25:02Z` (comment by `lebronjamesking`), which is more than 6 months before this run date `2026-07-12`. Therefore this issue is being closed under the scheduled stale-issue cleanup rule.

Evidence checked: Latest non-self public reply/reference: `2025-08-22T07:25:02Z` (comment by `lebronjamesking`).
Comments API: https://api.github.com/repos/verl-project/verl/issues/2708/comments
Timeline API: https://api.github.com/repos/verl-project/verl/issues/2708/timeline

If the author or community members can still provide a fresh reproduction, evidence, or requirement context, please comment here to request reopening, or open a new issue and link back to this one.

This response was generated by a scheduled Codex GPT5.5 issue processing task.
